File: src/lib/watermark/magazine.ts
import { normalizeModel } from "@/lib/exif/brand";
import type { ExifData, FrameParams, TemplateConfig } from "@/stores/types";
import {
  averageLuminance,
  backgroundFill,
  buildPreviewRenderPlan,
  cleanDisplayText,
  formatTakenAt,
  getPreviewFontFamily,
  resolvePreviewGeometryMetrics,
  resolveReadableTextAndDivider,
  sanitizeCustomLines,
} from "./classic-bottom";
import { WATERMARK_LAYOUT_SPEC } from "./layout-spec";

type PreviewData = {
  width: number;
  height: number;
  src: string;
  exif: ExifData;
};

type MagazineColumns = {
  left: string[];
  right: string[];
};

const COLUMN_GAP_PX = 24;
const COLUMN_INNER_PADDING_PX = 16;

export function buildMagazineColumns(
  exif: ExifData,
  config: TemplateConfig,
): MagazineColumns {
  const left: string[] = [];
  const right: string[] = [];

  if (config.showCamera) {
    const camera = normalizeModel(exif.camera.make, exif.camera.model);
    if (camera) left.push(camera);
  }
  if (config.showLens) {
    const lens = cleanDisplayText(exif.lens);
    if (lens) left.push(lens);
  }

  if (config.showParams) {
    if (exif.focalLength) right.push(`${Math.round(exif.focalLength)}mm`);
    if (exif.aperture) right.push(`f/${trimNumeric(exif.aperture)}`);
    if (exif.shutterSpeed) right.push(exif.shutterSpeed);
    if (exif.iso) right.push(`ISO ${exif.iso}`);
  }

  if (config.showDate) {
    const date = formatTakenAt(exif.takenAt, config.dateFormat);
    if (date) right.push(date);
  }

  for (const custom of sanitizeCustomLines(config.customLines)) {
    left.push(custom);
  }

  return { left, right };
}

export function drawMagazinePreview(
  canvas: HTMLCanvasElement,
  image: HTMLImageElement,
  logoImage: HTMLImageElement | null,
  photo: PreviewData,
  frameParams: FrameParams,
  config: TemplateConfig,
) {
  const ctx = canvas.getContext("2d");
  if (!ctx) return;

  const dpr = window.devicePixelRatio || 1;
  const baseWidth = 900;
  const watermarkActive = config.showWatermark ?? true;
  const columns = watermarkActive
    ? buildMagazineColumns(photo.exif, config)
    : { left: [], right: [] };
  const effectiveLogo = watermarkActive && config.showLogo ? logoImage : null;

  const primaryFontSize = Math.max(
    WATERMARK_LAYOUT_SPEC.baseMinPrimaryFontSize,
    frameParams.fontSize * WATERMARK_LAYOUT_SPEC.primaryFontScale,
  );
  const secondaryFontSize = Math.max(
    WATERMARK_LAYOUT_SPEC.baseMinSecondaryFontSize,
    frameParams.fontSize * WATERMARK_LAYOUT_SPEC.secondaryFontScale,
  );

  // Magazine bar shows a fixed two-row block per column; reserve enough space.
  const totalTextHeight =
    primaryFontSize + (columns.left.length > 1 ? secondaryFontSize + 6 : 0);

  const plan = buildPreviewRenderPlan({
    photoWidth: photo.width,
    photoHeight: photo.height,
    frameParams,
    templateKind: "magazine",
    totalTextHeight,
    logoOnlyWatermark: false,
    showWatermark: watermarkActive,
    baseWidth,
  });
  const geometry = resolvePreviewGeometryMetrics(frameParams);

  canvas.width = Math.round(plan.canvasW * dpr);
  canvas.height = Math.round(plan.canvasH * dpr);
  canvas.style.aspectRatio = `${plan.canvasW} / ${plan.canvasH}`;
  ctx.setTransform(dpr, 0, 0, dpr, 0, 0);
  ctx.clearRect(0, 0, plan.canvasW, plan.canvasH);

  ctx.fillStyle = backgroundFill(frameParams);
  ctx.fillRect(0, 0, plan.canvasW, plan.canvasH);

  if (frameParams.shadow) {
    ctx.save();
    ctx.shadowColor = `rgba(17,24,39,${frameParams.shadowOpacity / 100})`;
    ctx.shadowBlur = frameParams.shadowBlur;
    ctx.shadowOffsetY = plan.photoH * (frameParams.shadowOffsetY / 100);
    ctx.fillStyle = "#000000";
    roundRect(ctx, plan.imageX, plan.imageY, plan.photoW, plan.photoH, geometry.innerRadius);
    ctx.fill();
    ctx.restore();
  }

  ctx.save();
  roundRect(ctx, plan.imageX, plan.imageY, plan.photoW, plan.photoH, geometry.innerRadius);
  ctx.clip();
  ctx.drawImage(image, plan.imageX, plan.imageY, plan.photoW, plan.photoH);
  ctx.restore();

  if (frameParams.photoBorder > 0 && frameParams.photoBorderStyle !== "none") {
    ctx.save();
    ctx.strokeStyle = frameParams.photoBorderColor || "#ffffff";
    ctx.lineWidth = geometry.photoBorder;
    if (frameParams.photoBorderStyle === "dashed") {
      const dash = Math.max(4, geometry.photoBorder * 2.4);
      ctx.setLineDash([dash, dash * 0.6]);
    }
    roundRect(ctx, plan.imageX, plan.imageY, plan.photoW, plan.photoH, geometry.innerRadius);
    ctx.stroke();
    ctx.restore();
  }

  if (!watermarkActive || plan.infoBarHeight <= 0) return;

  // Sample background luminance under the bar for auto-contrast.
  const sampleY = Math.max(plan.barTop, 0);
  const sampleH = Math.max(20, plan.canvasH - sampleY);
  const luma = averageLuminance(ctx, 0, sampleY, plan.canvasW, sampleH);
  const readable = resolveReadableTextAndDivider({
    autoTextContrast: frameParams.autoTextContrast,
    averageLuminance: luma,
    fallbackTextColor: frameParams.textColor,
    fallbackDividerColor: frameParams.dividerColor,
  });

  const barCenterY = plan.barTop + plan.infoBarHeight / 2;
  const dividerX = plan.canvasW / 2;
  const left = plan.horizontalMargin + COLUMN_INNER_PADDING_PX;
  const right = plan.canvasW - plan.horizontalMargin - COLUMN_INNER_PADDING_PX;
  const leftEnd = dividerX - COLUMN_GAP_PX / 2;
  const rightStart = dividerX + COLUMN_GAP_PX / 2;

  // Optional vertical divider (only when both columns have content).
  if (frameParams.dividerShow && columns.left.length > 0 && columns.right.length > 0) {
    const inset = Math.min(16, plan.infoBarHeight * 0.25);
    ctx.strokeStyle = readable.dividerColor;
    ctx.lineWidth = 1;
    ctx.beginPath();
    ctx.moveTo(dividerX, plan.barTop + inset);
    ctx.lineTo(dividerX, plan.barTop + plan.infoBarHeight - inset);
    ctx.stroke();
  }

  ctx.fillStyle = readable.textColor;
  ctx.textBaseline = "middle";

  drawLeftColumn({
    ctx,
    columns,
    logo: effectiveLogo,
    fontFamily: frameParams.fontFamily,
    primaryFontSize,
    secondaryFontSize,
    logoGap: frameParams.logoGap,
    barCenterY,
    leftX: left,
    leftEnd,
  });

  drawRightColumn({
    ctx,
    items: columns.right,
    fontFamily: frameParams.fontFamily,
    fontSize: secondaryFontSize,
    barCenterY,
    rightX: right,
    rightStart,
  });
}

function drawLeftColumn({
  ctx,
  columns,
  logo,
  fontFamily,
  primaryFontSize,
  secondaryFontSize,
  logoGap,
  barCenterY,
  leftX,
  leftEnd,
}: {
  ctx: CanvasRenderingContext2D;
  columns: MagazineColumns;
  logo: HTMLImageElement | null;
  fontFamily: FrameParams["fontFamily"];
  primaryFontSize: number;
  secondaryFontSize: number;
  logoGap: number;
  barCenterY: number;
  leftX: number;
  leftEnd: number;
}) {
  const lines = columns.left;
  ctx.textAlign = "left";

  if (lines.length === 0 && !logo) return;

  const primary = lines[0] ?? "";
  const secondary = lines[1];
  const lineGap = 6;

  const primaryFont = `700 ${primaryFontSize}px ${getPreviewFontFamily(fontFamily)}`;
  ctx.font = primaryFont;

  const logoBox = logo
    ? calcLogoBox(logo, primaryFontSize * (WATERMARK_LAYOUT_SPEC.logoVisualScale ?? 1.18))
    : null;
  const logoTextGap = logoBox && primary ? logoGap : 0;

  let cursorX = leftX;
  if (logoBox) {
    drawLogo(ctx, logo!, logoBox, cursorX, barCenterY - logoBox.height / 2);
    cursorX += logoBox.width + logoTextGap;
  }

  const remaining = Math.max(0, leftEnd - cursorX);
  const effectivePrimary = clipText(ctx, primary, remaining);

  // Primary line baseline aligns with logo's vertical center.
  const primaryY = secondary ? barCenterY - lineGap / 2 - primaryFontSize / 4 : barCenterY;
  if (effectivePrimary) {
    ctx.fillText(effectivePrimary, cursorX, primaryY);
  }

  if (secondary) {
    const secondaryY = barCenterY + secondaryFontSize / 2 + lineGap / 2;
    ctx.font = `400 ${secondaryFontSize}px ${getPreviewFontFamily(fontFamily)}`;
    const secondaryRemaining = Math.max(0, leftEnd - leftX);
    const effectiveSecondary = clipText(ctx, secondary, secondaryRemaining);
    if (effectiveSecondary) {
      ctx.fillText(effectiveSecondary, leftX, secondaryY);
    }
  }
}

function drawRightColumn({
  ctx,
  items,
  fontFamily,
  fontSize,
  barCenterY,
  rightX,
  rightStart,
}: {
  ctx: CanvasRenderingContext2D;
  items: string[];
  fontFamily: FrameParams["fontFamily"];
  fontSize: number;
  barCenterY: number;
  rightX: number;
  rightStart: number;
}) {
  if (items.length === 0) return;
  ctx.textAlign = "right";
  ctx.font = `600 ${fontSize}px ${getPreviewFontFamily(fontFamily)}`;

  // Lay out all params on a single line if they fit; otherwise split into 2 rows.
  const separator = "  ·  ";
  const single = items.join(separator);
  const singleWidth = ctx.measureText(single).width;
  const available = rightX - rightStart;

  if (singleWidth <= available) {
    ctx.fillText(single, rightX, barCenterY);
    return;
  }

  // Two-row fallback: split items roughly evenly.
  const mid = Math.ceil(items.length / 2);
  const row1 = items.slice(0, mid).join(separator);
  const row2 = items.slice(mid).join(separator);
  const lineGap = 6;
  const y1 = barCenterY - fontSize / 2 - lineGap / 2 + fontSize / 2;
  const y2 = barCenterY + fontSize / 2 + lineGap / 2 - fontSize / 2;
  ctx.fillText(clipText(ctx, row1, available), rightX, y1);
  ctx.fillText(clipText(ctx, row2, available), rightX, y2);
}

function clipText(ctx: CanvasRenderingContext2D, text: string, maxWidth: number) {
  if (!text || maxWidth <= 0) return "";
  if (ctx.measureText(text).width <= maxWidth) return text;
  const ellipsis = "…";
  let end = text.length - 1;
  while (end > 0 && ctx.measureText(text.slice(0, end) + ellipsis).width > maxWidth) {
    end -= 1;
  }
  return end > 0 ? text.slice(0, end) + ellipsis : ellipsis;
}

type LogoBox = { width: number; height: number };

function calcLogoBox(image: HTMLImageElement, targetHeight: number): LogoBox {
  const w = Math.max(1, image.naturalWidth);
  const h = Math.max(1, image.naturalHeight);
  const height = Math.max(12, targetHeight);
  return { width: height * (w / h), height };
}

function drawLogo(
  ctx: CanvasRenderingContext2D,
  image: HTMLImageElement,
  size: LogoBox,
  x: number,
  y: number,
) {
  ctx.drawImage(image, x, y, size.width, size.height);
}

function roundRect(
  ctx: CanvasRenderingContext2D,
  x: number,
  y: number,
  width: number,
  height: number,
  radius: number,
) {
  const r = Math.max(0, Math.min(radius, width / 2, height / 2));
  ctx.beginPath();
  ctx.moveTo(x + r, y);
  ctx.arcTo(x + width, y, x + width, y + height, r);
  ctx.arcTo(x + width, y + height, x, y + height, r);
  ctx.arcTo(x, y + height, x, y, r);
  ctx.arcTo(x, y, x + width, y, r);
  ctx.closePath();
}

function trimNumeric(value: number) {
  return Number.isInteger(value) ? String(value) : value.toFixed(1);
}

File: src-tauri/src/render/templates/magazine.rs
use image::{
    imageops::{overlay, resize, FilterType},
    DynamicImage, Rgba, RgbaImage,
};

use crate::render::classic_bottom::{
    apply_rounded_corners, average_luminance, build_render_lines, build_render_plan,
    clean_display_text, draw_rounded_rect_stroke, draw_soft_shadow, format_camera, format_date,
    load_logo_rgba, parse_color, resize_photo, resolve_readable_colors, ExportExif,
    ExportFrameParams, ExportTemplateConfig,
};
use crate::render::layout_spec::watermark_layout_spec;
use crate::render::text::TextRenderer;

pub(crate) fn compose_magazine(
    source: DynamicImage,
    frame: &ExportFrameParams,
    exif: &ExportExif,
    config: &ExportTemplateConfig,
    text: Option<&TextRenderer>,
) -> RgbaImage {
    let src_w = source.width();
    let src_h = source.height();
    let rs = src_w as f32 / 900.0;

    let (left_lines, right_items) = build_magazine_columns(exif, config);
    let logo = if config.show_watermark && config.show_logo {
        load_logo_rgba(frame, exif, config)
    } else {
        None
    };

    let flat_lines: Vec<String> = left_lines.iter().chain(right_items.iter()).cloned().collect();
    let render_lines_for_plan = if config.show_watermark {
        build_render_lines(flat_lines, logo.is_some())
    } else {
        Vec::new()
    };
    let plan = build_render_plan(
        src_w,
        src_h,
        frame,
        "magazine",
        &render_lines_for_plan,
        text,
        logo.as_ref().map(|img| img.width() as f32 / img.height().max(1) as f32),
        config.show_watermark,
    );

    let bg = parse_color(&frame.bg_color, &frame.background);
    let mut canvas = RgbaImage::from_pixel(plan.canvas_w, plan.canvas_h, bg);

    let mut photo = resize_photo(source, plan.photo_w, plan.photo_h);
    let r = ((frame.inner_radius as f32 * rs).round() as u32)
        .min(plan.photo_h / 2)
        .min(plan.photo_w / 2);
    apply_rounded_corners(&mut photo, r);

    if frame.shadow {
        let shadow_offset_y =
            ((plan.photo_h as f32) * (frame.shadow_offset_y.max(0.0) / 100.0)).round() as u32;
        draw_soft_shadow(
            &mut canvas,
            plan.image_left,
            plan.image_top,
            plan.photo_w,
            plan.photo_h,
            r,
            (frame.shadow_blur as f32 * rs).round() as u32,
            shadow_offset_y,
            frame.shadow_opacity.min(100),
        );
    }
    overlay(&mut canvas, &photo, i64::from(plan.image_left), i64::from(plan.image_top));

    if frame.photo_border > 0 && frame.photo_border_style != "none" {
        let border = (frame.photo_border as f32 * rs).round() as u32;
        let border_color = if frame.photo_border_color.is_empty() {
            Rgba([255, 255, 255, 255])
        } else {
            parse_color(&frame.photo_border_color, "white")
        };
        draw_rounded_rect_stroke(
            &mut canvas,
            plan.image_left,
            plan.image_top,
            plan.photo_w,
            plan.photo_h,
            r,
            border,
            border_color,
        );
    }

    if !config.show_watermark || plan.canvas_h <= plan.bar_top {
        return canvas;
    }

    let spec = watermark_layout_spec();
    let primary_pt = (frame.font_size as f32 * spec.primary_font_scale)
        .max(spec.base_min_primary_font_size) * rs;
    let secondary_pt = (frame.font_size as f32 * spec.secondary_font_scale)
        .max(spec.base_min_secondary_font_size) * rs;

    let text_color = parse_color(&frame.text_color, "white");
    let divider_color = parse_color(&frame.divider_color, "white");

    let bar_top = plan.bar_top;
    let avg_luma = average_luminance(&canvas, 0, bar_top, plan.canvas_w, plan.canvas_h - bar_top);
    let (text_color, divider_color) =
        resolve_readable_colors(frame.auto_text_contrast, avg_luma, text_color, divider_color);

    let side_margin = (plan.canvas_w as f32 * 0.044).round() as u32;
    let col_inner_pad = (16.0 * rs).round();
    let divider_x = plan.canvas_w / 2;
    let col_gap = (24.0 * rs).round();
    let left_x = side_margin as f32 + col_inner_pad;
    let left_end = divider_x as f32 - col_gap / 2.0;
    let right_start = divider_x as f32 + col_gap / 2.0;
    let right_x = plan.canvas_w as f32 - side_margin as f32 - col_inner_pad;
    let bar_center_y = bar_top as f32 + (plan.canvas_h - bar_top) as f32 / 2.0;

    if frame.divider_show && !left_lines.is_empty() && !right_items.is_empty() {
        let inset = ((plan.canvas_h - bar_top) as f32 * 0.22).round() as u32;
        for y in (bar_top + inset)..(plan.canvas_h.saturating_sub(inset)) {
            if divider_x < plan.canvas_w {
                canvas.put_pixel(divider_x, y, divider_color);
            }
        }
    }

    let primary = left_lines.first().cloned().unwrap_or_default();
    let secondary = left_lines.get(1).cloned();

    let logo_ratio = logo.as_ref().map(|img| img.width() as f32 / img.height().max(1) as f32);
    let logo_target_h = primary_pt * spec.logo_visual_scale;
    let logo_w = logo_ratio.map(|ratio| logo_target_h * ratio).unwrap_or(0.0);
    let logo_gap = if logo.is_some() && !primary.is_empty() {
        frame.logo_gap as f32 * rs
    } else {
        0.0
    };

    let line_gap = 6.0 * rs;
    let (primary_ascent, primary_line_h, secondary_ascent, secondary_line_h) = text
        .map(|r| {
            (
                r.ascent(primary_pt, true),
                r.line_height(primary_pt, true),
                r.ascent(secondary_pt, false),
                r.line_height(secondary_pt, false),
            )
        })
        .unwrap_or((primary_pt * 0.8, primary_pt, secondary_pt * 0.8, secondary_pt));

    let (primary_y, secondary_y_opt) = if secondary.is_some() {
        let total_h = primary_line_h + line_gap + secondary_line_h;
        let block_top = bar_center_y - total_h / 2.0;
        (block_top, Some(block_top + primary_line_h + line_gap))
    } else {
        (bar_center_y - primary_line_h / 2.0, None)
    };

    let primary_baseline = primary_y + primary_ascent;

    if let Some(logo_img) = &logo {
        let logo_draw_w = (logo_w.round().max(1.0)) as u32;
        let logo_draw_h = (logo_target_h.round().max(1.0)) as u32;
        let resized = resize(logo_img, logo_draw_w, logo_draw_h, FilterType::Lanczos3);
        let logo_x = left_x.round() as i64;
        let logo_y = (primary_baseline + primary_pt * spec.logo_baseline_offset_ratio - logo_target_h)
            .round() as i64;
        overlay(&mut canvas, &resized, logo_x, logo_y);
    }

    let text_left_x = left_x + logo_w + logo_gap;

    if let Some(rend) = text {
        if !primary.is_empty() {
            let avail = (left_end - text_left_x).max(0.0);
            let clipped = clip_text(rend, &primary, avail, primary_pt, true);
            rend.draw(&mut canvas, &clipped, text_left_x, primary_y, primary_pt, true, text_color);
        }
        if let (Some(sec), Some(sec_y)) = (&secondary, secondary_y_opt) {
            let avail = (right_x - left_x).max(0.0);
            let clipped = clip_text(rend, sec, avail, secondary_pt, false);
            rend.draw(&mut canvas, &clipped, left_x, sec_y, secondary_pt, false, text_color);
        }
    }

    if let Some(rend) = text {
        if !right_items.is_empty() {
            let sep = "  \u{00B7}  ";
            let single = right_items.join(sep);
            let available = right_x - right_start;
            let single_w = rend.measure(&single, secondary_pt, true);

            if single_w <= available {
                let y = bar_center_y - secondary_line_h / 2.0;
                rend.draw(&mut canvas, &single, right_x - single_w, y, secondary_pt, true, text_color);
            } else {
                let mid = right_items.len().div_ceil(2);
                let row1 = right_items[..mid].join(sep);
                let row2 = right_items[mid..].join(sep);
                let total_h = secondary_line_h * 2.0 + line_gap;
                let block_top = bar_center_y - total_h / 2.0;
                let row2_y = block_top + secondary_line_h + line_gap;
                let c1 = clip_text(rend, &row1, available, secondary_pt, true);
                let c2 = clip_text(rend, &row2, available, secondary_pt, true);
                let w1 = rend.measure(&c1, secondary_pt, true).min(available);
                let w2 = rend.measure(&c2, secondary_pt, true).min(available);
                rend.draw(&mut canvas, &c1, right_x - w1, block_top, secondary_pt, true, text_color);
                rend.draw(&mut canvas, &c2, right_x - w2, row2_y, secondary_pt, true, text_color);
            }
        }
    }

    let _ = secondary_ascent;

    canvas
}

fn build_magazine_columns(
    exif: &ExportExif,
    config: &ExportTemplateConfig,
) -> (Vec<String>, Vec<String>) {
    let mut left: Vec<String> = Vec::new();
    let mut right: Vec<String> = Vec::new();

    if config.show_camera {
        let cam = format_camera(exif);
        if !cam.is_empty() {
            left.push(cam);
        }
    }
    if config.show_lens {
        let lens = clean_display_text(&exif.lens);
        if !lens.is_empty() {
            left.push(lens);
        }
    }
    for line in &config.custom_lines {
        let t = line.trim().to_string();
        if !t.is_empty() {
            left.push(t);
        }
    }

    if config.show_params {
        let items: Vec<String> = [
            if exif.focal_length > 0.0 {
                format!("{}mm", exif.focal_length.round() as u32)
            } else {
                String::new()
            },
            if exif.aperture > 0.0 {
                format!("f/{:.1}", exif.aperture)
            } else {
                String::new()
            },
            exif.shutter_speed.clone(),
            if exif.iso > 0 {
                format!("ISO {}", exif.iso)
            } else {
                String::new()
            },
        ]
        .into_iter()
        .filter(|s| !s.is_empty())
        .collect();
        right.extend(items);
    }
    if config.show_date {
        let date = format_date(&exif.taken_at, &config.date_format);
        if !date.is_empty() {
            right.push(date);
        }
    }

    (left, right)
}

fn clip_text(r: &TextRenderer, text: &str, max_width: f32, size: f32, bold: bool) -> String {
    if max_width <= 0.0 || text.is_empty() {
        return String::new();
    }
    if r.measure(text, size, bold) <= max_width {
        return text.to_string();
    }
    let ellipsis = "\u{2026}";
    let mut end = text.chars().count();
    while end > 0 {
        let s: String = text.chars().take(end).collect();
        let candidate = format!("{}{}", s, ellipsis);
        if r.measure(&candidate, size, bold) <= max_width {
            return candidate;
        }
        end -= 1;
    }
    ellipsis.to_string()
}
